sockeye/inference.py

```diff
@@ -2026,16 +2036,57 @@ def hybrid_forward(self, F, scores, offset):
         :param offset: Array to add to the hypothesis indices for offsetting in batch decoding.
         :return: The row indices, column indices and values of the k smallest items in matrix.
         """
-        folded_scores = F.reshape(scores, shape=(self.batch_size, self.k * self.vocab_size))
+        folded_scores = F.reshape(scores, shape=(self.batch_size, -1))
+
         values, indices = F.topk(folded_scores, axis=1, k=self.k, ret_typ='both', is_ascend=True)
+
         indices = F.reshape(F.cast(indices, 'int32'), shape=(-1,))
-        unraveled = F.unravel_index(indices, shape=(self.batch_size * self.k, self.vocab_size))
+        unraveled = F.unravel_index(indices, shape=(-1, self.vocab_size))
+
         best_hyp_indices, best_word_indices = F.split(unraveled, axis=0, num_outputs=2, squeeze_axis=True)
```

Comment:
This is awesome, as we can remove `self.batch_size` and `self.vocab_size` again from this block right?
I wonder if we can even remove the use of `batch_size` for `self.offset` in the constructor somehow. Is the newly added `batch_indices` array in the calling code not equal to `self.offset`?

In general, having to pass in a constant shape parameter (batch size) into these Blocks is a nuisance that we have to live with so far. If we ever want to add dynamic batch sizes for translation, this would need to be changed anyway. Ideally, we would create these auxiliary masks/arrays based on the input to the block dynamically, but if that is too much overhead, passing them in as inputs and create them once outside the block is the better option.

Reply:
I guess you are right, I didn't see the use of `self.batch_size` when reshaping into `folded_scores`. Afaik, MXNet doesn't really have an op (yet?) that returns the shape of its input, which can then be used as an argument to another op (most shape arguments of operators are not symbols but int tuples).

Reply:
Yeah, I wish you could write something like:

    unraveled = F.unravel_index(indices, shape=scores)